# Cleaning Scrapped world population data

Here, I will be cleaning the [world population data scrapped from the worldmeter website](https://www.worldmeters.info/world-population/population-by-country/). 

## Summary of Results

Data was successfully cleaned and the columns were stored in their proper formats. The packages used were `pandas` and `numpy`.

## Data Cleaning

### Step 1: Import the required libraries

Here we will be using `pandas` and `numpy` so we import those

In [169]:
try: 
    import pandas as pd 
    import numpy as np
except ModuleNotFoundError:
    !pip install pandas
    !pip install numpy 
    
    import pandas as pd 
    import numpy as np

### Step 2: Read your data into the notebook

The data in question is stored in CSV format so here we use the `read_csv()` method in `pandas` to read our file in as a DataFrame

In [192]:
worldPopData = pd.read_csv('./results/world-population-data-worldmeter.csv')

print(worldPopData.head(5))

  Country (or dependency) Population (2020) Yearly Change  Net Change  \
0                   China     1,439,323,776        0.39 %   5,540,090   
1                   India     1,380,004,385        0.99 %  13,586,631   
2           United States       331,002,651        0.59 %   1,937,734   
3               Indonesia       273,523,615        1.07 %   2,898,047   
4                Pakistan       220,892,340        2.00 %   4,327,022   

  Density (P/Km²) Land Area (Km²) Migrants (net) Fert. Rate Med. Age  \
0             153       9,388,211       -348,399        1.7       38   
1             464       2,973,190       -532,687        2.2       28   
2              36       9,147,420        954,806        1.8       38   
3             151       1,811,570        -98,955        2.3       30   
4             287         770,880       -233,379        3.6       23   

  Urban Pop % World Share  
0        61 %     18.47 %  
1        35 %     17.70 %  
2        83 %      4.25 %  
3        56 %   

Then we examine the data using the `info()`, `head()` and `tail()` methods in pandas to get an overview of our data

In [193]:
print(worldPopData.tail(5))

    Country (or dependency) Population (2020) Yearly Change Net Change  \
230              Montserrat             4,992        0.06 %          3   
231        Falkland Islands             3,480        3.05 %        103   
232                    Niue             1,626        0.68 %         11   
233                 Tokelau             1,357        1.27 %         17   
234                Holy See               801        0.25 %          2   

    Density (P/Km²) Land Area (Km²) Migrants (net) Fert. Rate Med. Age  \
230              50             100            NaN       N.A.     N.A.   
231               0          12,170            NaN       N.A.     N.A.   
232               6             260            NaN       N.A.     N.A.   
233             136              10            NaN       N.A.     N.A.   
234           2,003               0            NaN       N.A.     N.A.   

    Urban Pop % World Share  
230        10 %      0.00 %  
231        66 %      0.00 %  
232        46 %     

In [194]:
worldPopData.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 235 entries, 0 to 234
Data columns (total 11 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   Country (or dependency)  235 non-null    object
 1   Population (2020)        235 non-null    object
 2   Yearly Change            235 non-null    object
 3   Net Change               235 non-null    object
 4   Density (P/Km²)          235 non-null    object
 5   Land Area (Km²)          235 non-null    object
 6   Migrants (net)           201 non-null    object
 7   Fert. Rate               235 non-null    object
 8   Med. Age                 235 non-null    object
 9   Urban Pop %              235 non-null    object
 10  World Share              235 non-null    object
dtypes: object(11)
memory usage: 20.3+ KB


### Step 3: Convert Dataframe dtype to appropriate one for data wrangling (can be skipped)

Looking at the cell above (`info()` cell), we notice that the dtype is `object`. We need to onvert this to another datatype for analysis. We can do this by using the `convert_dtypes()` method in `pandas`.

In [195]:
worldPopData = worldPopData.convert_dtypes()

worldPopData

Country (or dependency) Population (2020) Yearly Change  Net Change  \
0                     China     1,439,323,776        0.39 %   5,540,090   
1                     India     1,380,004,385        0.99 %  13,586,631   
2             United States       331,002,651        0.59 %   1,937,734   
3                 Indonesia       273,523,615        1.07 %   2,898,047   
4                  Pakistan       220,892,340        2.00 %   4,327,022   
..                      ...               ...           ...         ...   
230              Montserrat             4,992        0.06 %           3   
231        Falkland Islands             3,480        3.05 %         103   
232                    Niue             1,626        0.68 %          11   
233                 Tokelau             1,357        1.27 %          17   
234                Holy See               801        0.25 %           2   

    Density (P/Km²) Land Area (Km²) Migrants (net) Fert. Rate Med. Age  \
0               153       9,388,211       -348,399        1.7       38   
1               464       2,973,190       -532,687        2.2       28   
2                36       9,147,420        954,806        1.8       38   
3               151       1,811,570        -98,955        2.3       30   
4               287         770,880       -233,379        3.6       23   
..              ...             ...            ...        ...      ...   
230              50             100           <NA>       N.A.     N.A.   
231               0          12,170           <NA>       N.A.     N.A.   
232               6             260           <NA>       N.A.     N.A.   
233             136              10           <NA>       N.A.     N.A.   
234           2,003               0           <NA>       N.A.     N.A.   

    Urban Pop % World Share  
0          61 %     18.47 %  
1          35 %     17.70 %  
2          83 %      4.25 %  
3          56 %      3.51 %  
4          35 %      2.83 %  
..          ...         ...  
230        10 %      0.00 %  
231        66 %      0.00 %  
232        46 %      0.00 %  
233         0 %      0.00 %  
234        N.A.      0.00 %  

[235 rows x 11 columns]

Looking at the cell below after running the above cell, we see that the `dtype` is now `string` type. We should be able to perform our data cleaning operations now

In [196]:
worldPopData.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 235 entries, 0 to 234
Data columns (total 11 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   Country (or dependency)  235 non-null    string
 1   Population (2020)        235 non-null    string
 2   Yearly Change            235 non-null    string
 3   Net Change               235 non-null    string
 4   Density (P/Km²)          235 non-null    string
 5   Land Area (Km²)          235 non-null    string
 6   Migrants (net)           201 non-null    string
 7   Fert. Rate               235 non-null    string
 8   Med. Age                 235 non-null    string
 9   Urban Pop %              235 non-null    string
 10  World Share              235 non-null    string
dtypes: string(11)
memory usage: 20.3 KB


### Step 4: Cleaned the column names

Looking at the column names, I noticed that they may make it a bit difficult to work with the data so I decided to rename theme to more analysis-friendly names.

In [197]:
worldPopData.columns

Index(['Country (or dependency)', 'Population (2020)', 'Yearly Change',
       'Net Change', 'Density (P/Km²)', 'Land Area (Km²)', 'Migrants (net)',
       'Fert. Rate', 'Med. Age', 'Urban Pop %', 'World Share'],
      dtype='object')

In [201]:
worldPopData = worldPopData.rename(columns={
    'Country (or dependency)': 'country',
    'Population (2020)': 'population_2020',
    'Yearly Change': 'yearly_change',
    'Net Change': 'net_change',
    'Density (P/Km²)': 'density',
    'Land Area (Km²)': 'land_area',
    'Migrants (net)': 'net_migrants',
    'Fert. Rate': 'fertility_rate',
    'Med. Age': 'median_age',
    'Urban Pop %': 'urban_pop_%',
    'World Share': 'world_share'
})

worldPopData

country population_2020 yearly_change net_change density  \
0               China      1439323776         0.39     5540090     153   
1               India      1380004385         0.99    13586631     464   
2       United States       331002651         0.59     1937734      36   
3           Indonesia       273523615         1.07     2898047     151   
4            Pakistan       220892340         2.00     4327022     287   
..                ...             ...           ...        ...     ...   
230        Montserrat            4992         0.06           3      50   
231  Falkland Islands            3480         3.05         103       0   
232              Niue            1626         0.68          11       6   
233           Tokelau            1357         1.27          17     136   
234          Holy See             801         0.25           2    2003   

    land_area net_migrants fertility_rate median_age urban_pop_% world_share  
0     9388211      -348399            1.7         38         61       18.47   
1     2973190      -532687            2.2         28         35       17.70   
2     9147420       954806            1.8         38         83        4.25   
3     1811570       -98955            2.3         30         56        3.51   
4      770880      -233379            3.6         23         35        2.83   
..        ...          ...            ...        ...         ...         ...  
230       100         <NA>           N.A.       N.A.         10        0.00   
231     12170         <NA>           N.A.       N.A.         66        0.00   
232       260         <NA>           N.A.       N.A.         46        0.00   
233        10         <NA>           N.A.       N.A.          0        0.00   
234         0         <NA>           N.A.       N.A.        N.A.       0.00   

[235 rows x 11 columns]

Looking at the dataset above, we notice that the numerical data has some special characters attached (like `,` and `%`). We need to remove these to be able to properly work on the data in our analysis. We'll do that next.

### Step 5: We remove unwanted characters from our numerical values

We use the `replace()` method in `pandas` to replace each unwanted character with an empty string. 

**Note:**
* Make sure to set the `regex` prop in `pd.replace()` to True as shown in the cell below.

In [216]:
for column in worldPopData:
    worldPopData[column].replace(r'(,)', '', regex=True,inplace = True)
    worldPopData[column].replace(r'(%)', '', regex=True,inplace = True)

worldPopData.convert_dtypes()
worldPopData.tail()

country population_2020 yearly_change net_change density  \
230        Montserrat            4992         0.06           3      50   
231  Falkland Islands            3480         3.05         103       0   
232              Niue            1626         0.68          11       6   
233           Tokelau            1357         1.27          17     136   
234          Holy See             801         0.25           2    2003   

    land_area net_migrants fertility_rate median_age urban_pop_% world_share  
230       100         <NA>           N.A.       N.A.         10        0.00   
231     12170         <NA>           N.A.       N.A.         66        0.00   
232       260         <NA>           N.A.       N.A.         46        0.00   
233        10         <NA>           N.A.       N.A.          0        0.00   
234         0         <NA>           N.A.       N.A.        N.A.       0.00

### Step 6: Check for missing values in columns of the DataFrame

Handling missing data is a part of data cleaning, here I needed to count the values in each column. I used the `value_counts()` method in pandas to do this by looping through the entire dataset and printing the results

In [211]:
for column in worldPopData:
    print(worldPopData[column].value_counts(dropna=False))

China                      1
Egypt                      1
South Korea                1
United States              1
Indonesia                  1
                          ..
Saint Pierre & Miquelon    1
Montserrat                 1
Falkland Islands           1
Niue                       1
Holy See                   1
Name: country, Length: 235, dtype: Int64
1439323776    1
102334404     1
51269185      1
331002651     1
273523615     1
             ..
5794          1
4992          1
3480          1
1626          1
801           1
Name: population_2020, Length: 235, dtype: Int64
1.48      4
0.67      3
0.97      3
0.25      3
0.32      3
         ..
1.57      1
0.62      1
0.16      1
-0.36     1
1.27      1
Name: yearly_change, Length: 174, dtype: Int64
68         2
1464463    1
543448     1
43877      1
1937734    1
          ..
-28        1
3          1
103        1
11         1
2          1
Name: net_change, Length: 234, dtype: Int64
25      8
4       6
16      5
18      5
83      5

We can also chain the `isnull()` and `sum()` methods to achieve this same effect. However, I noticed that in the cell above, we found missing values in the `net_migrants`, `fertility_rate`, `median_age` and `urban_pop_%` columns but the cell below (chaining the `isnull()` and `sum()` methods) show us that there is only one column with missing values (which is `net_migrants`). 

This is likely because the source had filled in the value `N.A.` for some of the missing data and left the rest null. It's likely due to human error but this meant for the data indicated as `N.A.`, BeautifulSoup read that as the text `N.A` (like it was a `string` value rather than `null` like it should be).

We can still work with this, it just means we will have to filter with respect to `N.A.`when getting our null values for the `fertility_rate`, `median_age` and `urban_pop_%` columns

In [213]:
worldPopData.isnull().sum()

country             0
population_2020     0
yearly_change       0
net_change          0
density             0
land_area           0
net_migrants       34
fertility_rate      0
median_age          0
urban_pop_%         0
world_share         0
dtype: int64

### Step 7: Fill in the missing values

When handling missing values, you can either delete them or fill them depending on the amount of missing data. Ideally, its okay to delete missing rows or columns if they can be done without.

I chose to fill in the missing data because deleting would mean loss of quite a bit of valuable data. I am filling them with the mean values of their columns so I'll use the `describe()` method in `pandas` to get a statistical breakdown of the data for reference purposes.

In [248]:
worldPopData.describe()

population_2020  yearly_change    net_change       density  \
count     2.350000e+02     235.000000  2.350000e+02    235.000000   
mean      3.317120e+07       1.104298  3.461058e+05    478.408511   
std       1.351370e+08       1.075665  1.128255e+06   2331.282424   
min       8.010000e+02      -2.470000 -3.838400e+05      0.000000   
25%       4.188015e+05       0.320000  4.240000e+02     37.000000   
50%       5.459642e+06       0.970000  3.917000e+04     95.000000   
75%       2.057705e+07       1.850000  2.496600e+05    240.000000   
max       1.439324e+09       3.840000  1.358663e+07  26337.000000   

          land_area   net_migrants  fertility_rate  median_age  urban_pop_%  \
count  2.350000e+02     235.000000      235.000000  235.000000   235.000000   
mean   5.535918e+05      -5.442786        2.635689   30.621891    59.445946   
std    1.687796e+06  113983.547562        1.176840    8.431033    23.420793   
min    0.000000e+00 -653249.000000        1.100000   15.000000     0.000000   
25%    2.545000e+03   -8608.000000        1.800000   24.000000    43.000000   
50%    7.724000e+04      -5.442786        2.302553   30.621891    59.445946   
75%    4.038200e+05    4900.000000        3.050000   37.500000    78.500000   
max    1.637687e+07  954806.000000        7.000000   48.000000   100.000000   

       world_share  
count   235.000000  
mean      0.425489  
std       1.733713  
min       0.000000  
25%       0.010000  
50%       0.070000  
75%       0.265000  
max      18.470000

So now we fill in the missing values by calculating the mean.

**Note:** 
* Convert the `dtype` of your columns to `float` or `int` before working. You can't perform math operations on string data.

In [227]:
net_migrants = worldPopData['net_migrants'].astype(float)
not_null_migrants = net_migrants[net_migrants.notnull()]
net_migrants[net_migrants.isnull()] = not_null_migrants.mean()
worldPopData['net_migrants'] = worldPopData['net_migrants'].astype(float)
worldPopData['net_migrants'] = net_migrants

worldPopData['net_migrants']

0     -348399.000000
1     -532687.000000
2      954806.000000
3      -98955.000000
4     -233379.000000
           ...      
230        -5.442786
231        -5.442786
232        -5.442786
233        -5.442786
234        -5.442786
Name: net_migrants, Length: 235, dtype: float64

In [241]:
for column in worldPopData[['fertility_rate', 'median_age', 'urban_pop_%']]:
    data_series = worldPopData[column]
    not_null_data = data_series[data_series != 'N.A.'].astype(float)
    data_series[data_series == 'N.A.'] = not_null_data.mean().astype(str)
    worldPopData[column] = data_series.astype(float)
    
worldPopData[['fertility_rate', 'median_age', 'urban_pop_%']].tail()

C:\Users\DEWA\AppData\Local\Temp/ipykernel_12548/3985628130.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_series[data_series == 'N.A.'] = not_null_data.mean().astype(str)


fertility_rate  median_age  urban_pop_%
230        2.302553   30.621891    10.000000
231        2.302553   30.621891    66.000000
232        2.302553   30.621891    46.000000
233        2.302553   30.621891     0.000000
234        2.302553   30.621891    59.445946

Taking a look at the DataFrame's info again, we see that all null values are filled because we see the Non-null Count column shows all columns have 235 rows of data as expected

In [242]:
worldPopData.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 235 entries, 0 to 234
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   country          235 non-null    string 
 1   population_2020  235 non-null    string 
 2   yearly_change    235 non-null    string 
 3   net_change       235 non-null    string 
 4   density          235 non-null    string 
 5   land_area        235 non-null    string 
 6   net_migrants     235 non-null    float64
 7   fertility_rate   235 non-null    float64
 8   median_age       235 non-null    float64
 9   urban_pop_%      235 non-null    float64
 10  world_share      235 non-null    string 
dtypes: float64(4), string(7)
memory usage: 20.3 KB


### Step 8: Convert the dtypes of all columns to the proper types

Remember we have most of our columns as strings (see result from `info()` method above). Some of the data is actually numerical and as such should not be stored as `string`. `String` is for categorical data (labels) while we use `int` and `float` for numerical data. 

We will use the `astype()` method to change the dtypes to the proper types

In [243]:
for column in worldPopData[['population_2020', 'net_change', 'density', 'land_area']]:
    worldPopData[column] = worldPopData[column].astype(int)

In [244]:
for column in worldPopData[['yearly_change', 'world_share']]:
    worldPopData[column] = worldPopData[column].astype(float) 

Now lets look at the DataFrame info below, we can see that the dtype is now much more suited to the data

In [246]:
worldPopData.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 235 entries, 0 to 234
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   country          235 non-null    string 
 1   population_2020  235 non-null    int32  
 2   yearly_change    235 non-null    float64
 3   net_change       235 non-null    int32  
 4   density          235 non-null    int32  
 5   land_area        235 non-null    int32  
 6   net_migrants     235 non-null    float64
 7   fertility_rate   235 non-null    float64
 8   median_age       235 non-null    float64
 9   urban_pop_%      235 non-null    float64
 10  world_share      235 non-null    float64
dtypes: float64(6), int32(4), string(1)
memory usage: 16.6 KB


### Step 9: Save the cleaned data to a file

Now that our data has been cleaned, we can store it in the file format we wish. I stored this as a CSV file so i used `pandas`' `to_csv()` method to do this

In [249]:
worldPopData.to_csv('./results/world-population-data-worldmeter-cleaned.csv', index=False)

## Conclusion

Data was successfully cleaned and the columns were stored in their proper formats. The packages used were `pandas` and `numpy`.